Add batch optimiser for symbolic expressions

jppelteret opened the pull request.

jppelteret commented:
/rebuild

Rombur commented on tests/symengine/batch_optimizer_01_1a.cc:
```diff
@@ -0,0 +1,41 @@
+// Test that all of the low-level math operations function as expected
+// and that their values and derivatives can be computed using the
+// optimizer.
+// For this test the optimizer is in mode "off", i.e. plain old
```
There is no `off` mode anymore

jppelteret replied:
Thanks, I've fixed this up.

Rombur commented on tests/symengine/batch_optimizer_01_1b.cc:
```diff
@@ -0,0 +1,42 @@
+// Test that all of the low-level math operations function as expected
+// and that their values and derivatives can be computed using the
+// optimizer.
+// For this test the optimizer is in mode "off", i.e. plain old
```
There is no more `off` mode anymore.

jppelteret commented:
> I trust you that it's the same thing I've already reviewed

Indeed, it is. I just squashed the fixes back into the main commits.

> You need to add an entry in the changelog.

Thanks for the reminder! I've now added one.

Rombur commented:
I'll merge this PR tomorrow unless someone as an objection (you have 24 hours to review 25 thousands line so don't forget the Red Bull)

jppelteret commented:
Alternatively, the main class documentation and the doxygen module are
relatively light reading :-D

On Wed, 06 May 2020, 00:16 Bruno Turcksin, <notifications@github.com> wrote:

> I'll merge this PR tomorrow unless someone as an objection (you have 24
> hours to review 25 thousands line so don't forget the Red Bull)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dealii/dealii/pull/10028#issuecomment-624335305>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACPZ3QWACUN2MTHUW6QV3Z3RQCFZ7ANCNFSM4MYRVD7A>
> .
>

bangerth commented:
Thanks, @jppelteret and @Rombur for working through this all!

I think we really ought to make writing a tutorial for all of this a priority. step-33 just isn't very good in this regard. Maybe an adaptation of step-15 might work?

jppelteret commented:
> I think we really ought to make writing a tutorial for all of this a priority.

The plan was to write an introductory tutorial at the hackathon. Let's see if I can still manage that. There's also the [code gallery example for AD](https://github.com/dealii/code-gallery/tree/master/Quasi_static_Finite_strain_Compressible_Elasticity), and in particular [its extension](https://github.com/dealii/code-gallery/pull/50). I think I'll just touch and an merge that PR, and at least we can reference it for now.
